### PR TITLE
fix(elbv2): unable to deploy template with IPv4 load balancer when denyAllIgwTraffic set

### LINKED ISSUE #30247: (aws-elasticloadbalancingv2): Prevent using `denyAllIgwTraffic` for load balancers that do not use dual stack addressing
### Describe the feature

Setting `denyAllIgwTraffic` flag for Load balancers with IPV4 addressing is not allowed, and cause this deployment error  `Load balancer attribute key 'ipv6.deny_all_igw_traffic' is not supported on load balancers with IP address type 'ipv4'.`

Sample usecase:
```
new elbv2.NetworkLoadBalancer(stack, 'NLB', {
  vpc,
  crossZoneEnabled: true,
  deletionProtection: false,
  denyAllIgwTraffic: true,
  clientRoutingPolicy: elbv2.ClientRoutingPolicy.PARTIAL_AVAILABILITY_ZONE_AFFINITY,
});
```

Also, there are some integration test cases that could not be deployed because of this issue.

### Use Case

It is better to figure out this issue during synthesize time instead of deployment time.

### Proposed Solution

Prevent customers form setting `denyAllIgwTraffic` flag is the Load balancers do not use dual stack addressing.

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

v2.142.0

### Environment details (OS name and version, etc.)

macos

### PR DESCRIPTION
**Pull request by Michae1CC**
### Issue # (if applicable)

Closes #30247 .

### Reason for this change

Integ test for NLB attributes ([integ.nlb-attributes.ts](https://github.com/aws/aws-cdk/blob/4f1c94b27ef7f4ceccea0ff39625c0e8add31c9f/packages/%40aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb-attributes.ts)) fails to deploy due to an error. The error occurs when `denyAllIgwTraffic` is explicitly set for load balancers with Ipv4 addressing, the `ipv6.deny_all_igw_traffic` attribute is set.

### Description of changes

- Remove the denyAllIgwTraffic setting from integ.nlb-attribute.ts
- Instead, set denyAllIgwTraffic in integ.nlb.dualstack.internal.ts.
- Raise an error during synthesis if `denyAllIgwTraffic` is set on a load balancer that does not use dual stack addressing.

### Description of how you validated changes

- Added new unit tests for different combinations of `denyAllIgwTraffic` and `ipAddressType`
- Updated existing integration test

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-elasticloadbalancingv2/test/nlb/load-balancer.test.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-elasticloadbalancingv2/test/nlb/load-balancer.test.ts:487]:
```suggestion
  test('throw error for denyAllIgwTraffic set to false for Ipv4 adressing', () => {
```

File: packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts:254]: `ipAddressType` is meant to be `IpAddressType.IPV4` by default, and `denyAllIgwTraffic` is supposed to be `false`, I assume this condition is incorrect?

File: packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json
Comment (nmussy) [inline packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: That's a scary change, I'm going to assume it's an unintended side effect. Given the integration was being deployed before your change, I assume the integ stack was being deployed successfully

File: packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts
Comment (Michae1CC) [reply packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts:254]: Appreciate the feedback! Apologies for the spiel

Yes, `ipAddressType` is meant to be `IpAddressType.IPV4`. As for `denyAllIgwTraffic` the docs mention [_"optional, default: false for internet-facing load balancers and true for internal load balancers"_](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2.ApplicationLoadBalancer.html#denyalligwtraffic). Looking at code though, `denyAllIgwTraffic` is never actually given a default value which is why we have this check for `undefined` here.

The `denyAllIgwTraffic` was created from this issue: https://github.com/aws/aws-cdk/issues/29520. Basically 
`denyAllIgwTraffic` is supposed to set `ipv6.deny_all_igw_traffic`. However, having `ipv6.deny_all_igw_traffic` set (to either true or false) for _any_ Ipv4 load balancer will cause a failed deployment. I'm starting to think the condition should actually be this:

```suggestion
    if (additionalProps.ipAddressType === IpAddressType.IPV4 && baseProps.denyAllIgwTraffic !== undefined) {
```

File: packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json
Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: Hehe, well that's the thing, I _could not_ deploy the integration test before my change. The deployment failed with
```txt
Resource handler returned message: "Load balancer attribute key 'ipv6.deny_all_igw_traffic' is not supported on load balancers with IP address type 'ipv4'.
```
That's why I renamed the test to, in effect, remove the old test and replace with a template you can deploy (the name I use also aligns with the existing alb test). This particular integration test was added very recently. I can only think that who ever created it only synthesized the test and didn't try deploying.

**Comment by nmussy** [comment]
@Michae1CC I also left you a general comment in the review, in case you didn't see it: https://github.com/aws/aws-cdk/pull/29956#pullrequestreview-2022389300

**Comment by Michae1CC** [comment]
> @Michae1CC I also left you a general comment in the review, in case you didn't see it: [#29956 (review)](https://github.com/aws/aws-cdk/pull/29956#pullrequestreview-2022389300)

Yes, thanks again! I'll add more tests once we've aligned on the expected behaviour.

File: packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts:254]: > Looking at code though, `denyAllIgwTraffic` is never actually given a default value which is why we have this check for undefined here.

The CDK doesn't always set the default values, in fact most of the time CloudFormation does. This is the case here, although it's more complicated than just defaulted to `false`, see [docs](https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_LoadBalancerAttribute.html). The `@default` value needs to be updated to reflect this

File: packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json
Comment (nmussy) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: @badmintoncryer Could we get some clarification on this? See #29521

Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: Surely I haven't created a snapshot without deploying it... 
I'll try it later. Please wait for a while.

Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: @nmussy @Michae1CC 
I tried to deploy `integ.nlb-attributes.ts` but it cannot. I'm really sorry and I don't know how I could create snapshot files...

I have conducted deployment verification for internal ALBs and NLBs across several patterns.
The results are the same for both ALB and NLB.

|`ipAddressType`|`denyAllIgwTraffic`|deployment result|error message|
|-|-|-|-|
|IPV4|true|failed|Load balancer attribute key 'ipv6.deny_all_igw_traffic' is not supported on load balancers with IP address type 'ipv4'.|
|IPV4|false|failed|Load balancer attribute key 'ipv6.deny_all_igw_traffic' is not supported on load balancers with IP address type 'ipv4'.|
|DUAL_STACK|true|success||
|DUAL_STACK|false|failed|Load balancer attribute key 'ipv6.deny_all_igw_traffic' cannot be modified for load balancers of type 'network'. |

Based on these results, I propose the following amendments:

1. If `denyAllIgwTraffic` is defined, return an error if `ipAddressType` is not `DUAL_STACK`.
2. Revise the existing `integ.nlb-attributes.ts` to make it deployable:
- Remove the `denyAllIgwTraffic` setting.
- Instead, set `denyAllIgwTraffic` in `integ.nlb.dualstack.internal.ts`.

For item 2, it might be better to handle it as a separate issue. 
If so, I will take responsibility for addressing it.

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: Thanks for this @badmintoncryer , I agree with you on both points. I'll address your first point here and leave you to fix the integration test. @nmussy I'll create the additional unit tests once I've finished making these changes.

**Comment by Michae1CC** [comment]
Exemption Request, @badmintoncryer will amend the integration test/s in a separate issue.

File: packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json
Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: @Michae1CC I will handle the resolution of the integration test. Although I have not dealt with this case before, I believe it would be best to wait for the merge of the integration test modification PR before moving this PR to in review.

@nmussy What are your thoughts on this?

Comment (nmussy) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: I don't see the need or point of doing two PRs to accomplish this. We won't be able to merge the fix without updating the integ and its snapshot, and just merging an update of the snapshot seems pointless, it doesn't resolve any issue for CDK users. I'm not really sure who should take precedence in this case, Kazuho should be responsible for this change, but this is Michael's first PR and he's already started working on it. 

I also don't think it's a high priority bug, given that CloudFormation is already catching this issue, this is just not great DX. I'll let you guys decide but I don't think two PRs is the way to go 👍

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: That's a good point, it seems strange to leave the integration testing broken when we've made a fix for it here. Happy to also fix the integration tests ie, revise the `integ.nlb-attributes.ts` test and add update the `integ.nlb.dualstack.internal.ts` test. How does that sound?

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: @badmintoncryer just saw your new PR. I don't mind either way.

Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: I apologize, I just noticed this comment. I will close my PR, could you please incorporate the same content?

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: Sure, I'll update this PR and handle integration tests 👍

Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: Thank you very much. I apologize for any inconvenience caused.

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json]: No problem 😄

**Comment by Michae1CC** [comment]
Hey both, updated the integration tests. Let me know what you think.

File: packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json
Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: @Michae1CC My only suggestion is that [the PR title should describe the content of the bug being resolved](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request), rather than just mentioning the resolution of integration test errors.

```md
Titles for feat and fix PRs end up in the change log. Think about what makes most sense for users reading the changelog while writing them.
- feat: describe the feature (not the action of creating the commit or PR, for example, avoid words like "added" or "changed")
- fix: describe the bug (not the solution)
```

Since I am unable to conduct community reviews, it would be best for the rest to be checked by @nmussy .

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: Gotcha, maybe something like `integ test for NLB attributes failing to deploy due to setting denyAllIgwTraffic`?

Comment (badmintoncryer) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: I apologize, I finally realized that this was actually a PR intended to resolve an integration test error. My English skills are limited and I made a significant misunderstanding. My suggestion to split the PR also made no sense. I'm sorry.

Without specifically mentioning the integration test, how about using the title: 
'fix(elbv2): unable to deploy an IPv4 load balancer with `denyAllIgwTraffic` enabled'?
or considering it as the addition of a verification feature,
'feat(elbv2): verification of `denyAllIgwTraffic` for IPv4 Load Balancers'

However, I would appreciate it if you could consider nmussy opinion on this matter as well.

Comment (nmussy) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: No need to apologize! And I agree, this PR's goal is not to fix an integration test, it's to prevent this pattern from being synthesized in the first place

Comment (Michae1CC) [reply packages/@aws-cdk-testing/framework-integ/test/aws-elasticloadbalancingv2/test/integ.nlb.attributes.js.snapshot/aws-cdk-nlb-attributes-integ.template.json:409]: Thanks both, updated the title

**Comment by Michae1CC** [comment]
@nmussy just floating this back up, are there any other changes you wanted? @badmintoncryer seems happy with it, but mentioned they would leave the review to you since they're not part of the cdk team.

**Review by nmussy — APPROVED** [review]

**Comment by jrobbins-LiveData** [comment]
I don't know if the problem I am having trying to deploy an ELBv2 with an IPv4 VPC is related to this or not (sorry if irrelevant), but I am getting this error

```
AlbCdk-test | 14/22 | 8:28:08 PM | CREATE_FAILED        | AWS::ElasticLoadBalancingV2::LoadBalancer | ALB (ALBAEE750D2) Resource handler returned message: "Load balancer attribute key 'ipv6.deny_all_igw_traffic' is not supported on load balancers with IP address type 'ipv4'. (Service: ElasticLoadBalancingV2, Status Code: 400, Request ID: cfbc6c78-d50c-4760-a163-020bceffe4d2)" (RequestToken: c487d145-cdbc-6431-112b-daa1a9bc652f, HandlerErrorCode: InvalidRequest)
```

The Python CDK code is

```
        alb = elbv2.ApplicationLoadBalancer(
            self, 'ALB',
            internet_facing=True,
            http2_enabled=True,
            ip_address_type=elbv2.IpAddressType.IPV4,
            drop_invalid_header_fields=True,
            desync_mitigation_mode=elbv2.DesyncMitigationMode.DEFENSIVE,
            client_keep_alive=cdk.Duration.seconds(500),
            deny_all_igw_traffic=False,
            preserve_host_header=True,
            x_amzn_tls_version_and_cipher_suite_headers=True,
            preserve_xff_client_port=True,
            xff_header_processing_mode=elbv2.XffHeaderProcessingMode.APPEND,
            waf_fail_open=True,
            vpc=vpc,
            vpc_subnets=vpc_subnets
            # security_group=security_group
        )
```

The VPC is IPv4 only. 

The subnets are the 3 public subnets of that VPC.

**Comment by Michae1CC** [comment]
> I don't know if the problem I am having trying to deploy an ELBv2 with an IPv4 VPC is related to this issue or not (sorry if irrelevant), but I am getting this error
> 
> ```
> AlbCdk-test | 14/22 | 8:28:08 PM | CREATE_FAILED        | AWS::ElasticLoadBalancingV2::LoadBalancer | ALB (ALBAEE750D2) Resource handler returned message: "Load balancer attribute key 'ipv6.deny_all_igw_traffic' is not supported on load balancers with IP address type 'ipv4'. (Service: ElasticLoadBalancingV2, Status Code: 400, Request ID: cfbc6c78-d50c-4760-a163-020bceffe4d2)" (RequestToken: c487d145-cdbc-6431-112b-daa1a9bc652f, HandlerErrorCode: InvalidRequest)
> ```
> 
> The Python CDK code is
> 
> ```
>         alb = elbv2.ApplicationLoadBalancer(
>             self, 'ALB',
>             internet_facing=True,
>             http2_enabled=True,
>             ip_address_type=elbv2.IpAddressType.IPV4,
>             drop_invalid_header_fields=True,
>             desync_mitigation_mode=elbv2.DesyncMitigationMode.DEFENSIVE,
>             client_keep_alive=cdk.Duration.seconds(500),
>             deny_all_igw_traffic=False,
>             preserve_host_header=True,
>             x_amzn_tls_version_and_cipher_suite_headers=True,
>             preserve_xff_client_port=True,
>             xff_header_processing_mode=elbv2.XffHeaderProcessingMode.APPEND,
>             waf_fail_open=True,
>             vpc=vpc,
>             vpc_subnets=vpc_subnets
>             # security_group=security_group
>         )
> ```
> 
> The VPC is IPv4 only.
> 
> The subnets are the 3 public subnets of that VPC.

@jrobbins-LiveData yes, this problem is directly related to these changes. Currently, having `deny_all_igw_traffic` set when using `elbv2.IpAddressType.IPV4` will cause the template to fail during deploy. Part of these changes will catch this by throwing an error during synthesis.

**Comment by Michae1CC** [comment]
@paulhcsun , @GavinZZ would either of you mind reviewing this one - thanks.

**Comment by jrobbins-LiveData** [comment]
@Michae1CC, I commented out `deny_all_igw_traffic=False` and the deployment worked.

```
        alb = elbv2.ApplicationLoadBalancer(
            self, 'ALB',
            internet_facing=True,
            http2_enabled=True,
            ip_address_type=elbv2.IpAddressType.IPV4,
            drop_invalid_header_fields=True,
            desync_mitigation_mode=elbv2.DesyncMitigationMode.DEFENSIVE,
            client_keep_alive=cdk.Duration.seconds(500),
            # deny_all_igw_traffic=False,
            preserve_host_header=True,
            x_amzn_tls_version_and_cipher_suite_headers=True,
            preserve_xff_client_port=True,
            xff_header_processing_mode=elbv2.XffHeaderProcessingMode.APPEND,
            waf_fail_open=True,
            vpc=vpc,
            vpc_subnets=vpc_subnets
            # security_group=security_group
        )
```

The [CDK help](https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_elasticloadbalancingv2/ApplicationLoadBalancerProps.html) indeed does not list a default for this param as it is marked as `Optional`.

> deny_all_igw_traffic (Optional[bool]) – Indicates whether the load balancer blocks traffic through the Internet Gateway (IGW). Default: - false for internet-facing load balancers and true for internal load balancers

Does that mean that `deny_all_igw_traffic` is a tri-state param, or is that what you are changing?  Just curious as to what the new behavior will be post-PR?

**Comment by Michae1CC** [comment]
> Just curious as to what the new behavior will be post-PR?

This PR isn't changing much. The biggest change (in terms of functionality) will be that the deploy error you got will now occur during synthesis with a more descriptive message.

**Comment by Michae1CC** [comment]
@moelasmar , let me know if anything else needs to change - thanks.

**Comment by moelasmar** [comment]
thanks @Michae1CC for raising this PR and fix this issue. It looks good to me, I will create a github issue for this PR, and link it to this PR (just for tracking purposes). I need also to verify the integration test cases on my side before approving it.

**Review by moelasmar — APPROVED** [review]

**Review by moelasmar — APPROVED** [review]

**Pull request merged**
